src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -331,6 +334,74 @@ struct ScalarOp<ONNXTanOp> {
   using IOp = NotSuportedScalarOp;
 };
 
+//===----------------------------------------------------------------------===//
+// Scalar unary ops for lowering ONNXGeluOp
+//===----------------------------------------------------------------------===//
+template <>
+struct ScalarOp<ONNXGeluOp> {
+  using FOp = CustomScalarOp;
+  using IOp = CustomScalarOp;
+};
+
+template <>
+double analyzeSimdFor<ONNXGeluOp>(
+    Type t, Operation *op, int64_t &von, int64_t &son) {
+  double results;
+  StringRef approximate = dyn_cast<ONNXGeluOp>(op).getApproximate();
+  if (approximate == "none")
+    results = simdAnalysis({GenericOps::ArithmeticGop, GenericOps::DivGop,
+                               GenericOps::MulGop, GenericOps::SqrtGop},
+        {1, 1, 2, 1}, t, von, son);
+  if (approximate == "tanh")
+    results = simdAnalysis(
+        {GenericOps::ArithmeticGop, GenericOps::DivGop, GenericOps::MulGop,
+            GenericOps::PowGop, GenericOps::SqrtGop},
+        {2, 1, 4, 1, 1}, t, von, son);
+  return results;
+}
+
+template <>
+Value emitScalarOpFor<ONNXGeluOp>(ConversionPatternRewriter &rewriter,
+    Location loc, Operation *op, Type elementType,
+    ArrayRef<Value> scalarOperands) {
+  Value operand = scalarOperands[0];
+
+  CheckIfCustomScalarOpIsSupported<ONNXGeluOp>(elementType);
+  MultiDialectBuilder<KrnlBuilder, MathBuilder> create(rewriter, loc);
+
+  // Approximate is a required attribute and should have a default value of
+  // "none". "approximate = none" simply implies no approximation will take
+  // place. However, "approximate" can also have a string value of "tanh" which
+  // indicates the use of tanh approximation.
+  StringRef approximate = dyn_cast<ONNXGeluOp>(op).getApproximate();
+
+  // Create constants
+  Value half = create.math.constant(elementType, 0.5);
+  Value one = create.math.constant(elementType, 1);
+  Value three = create.math.constant(elementType, 3);
+  Value decimal = create.math.constant(elementType, 0.044715);
+  Value sqrt_two_over_pi = create.math.constant(elementType, sqrt(2 / M_PI));
+  Value sqrt_two = create.math.constant(elementType, sqrt(2));
+
+  // Calculations
+  Value div = create.math.div(operand, sqrt_two);
+  Value erfApprox = create.math.erf(div);
+  Value add = create.math.add(one, erfApprox);
+  Value dec = create.math.add(
+      operand, create.math.mul(decimal, create.math.pow(operand, three)));
+  Value tanhApprox = create.math.tanh(create.math.mul(sqrt_two_over_pi, dec));
+  // Approximate = none returns an output of y = 0.5 * x * (1 +
+  // erf(x/sqrt(2)))
+  if (approximate == "none")
+    return create.math.mul(create.math.mul(half, operand), add);
+  // Approximate = tanh returns an output of y = 0.5 * x * (1 + Tanh(sqrt(2/pi)
+  // * (x + 0.044715 * x^3)))
+  if (approximate == "tanh")
+    return create.math.mul(
+        create.math.mul(half, operand), create.math.add(one, tanhApprox));
```

Comment:
I would create a global variable for `create.math.mul(half, operand)` since it's used in both cases.

Reply:
kk done